core/src/main/scala/kafka/server/AdminManager.scala

```diff
@@ -194,6 +197,64 @@ class AdminManager(val config: KafkaConfig,
     }
   }
 
+  def createPartitions(timeout: Int,
+                       newPartitions: Map[String, NewPartitions],
+                       validateOnly: Boolean,
+                       listenerName: ListenerName,
+                       callback: Map[String, ApiError] => Unit): Unit = {
+    var waiting = Map.empty[String, NewPartitions]
+    var done = Map.empty[String, ApiError]
+
+    newPartitions.foreach{case (topic, newPartition: NewPartitions) =>
+      try {
+        val oldNumPartitions = zkUtils.getTopicPartitionCount(topic) match {
+          case Some(count) => count
+          case None => throw new InvalidTopicException()
+        }
+        val newNumPartitions = newPartition.totalCount
+        val numPartitionsIncrement = newNumPartitions - oldNumPartitions
+        if (numPartitionsIncrement <= 0) {
+          throw new InvalidPartitionsException(s"Topic currently has $oldNumPartitions partitions: Cannot create $numPartitionsIncrement partitions.")
+        }
+        val reassignmentStr = if (newPartition.assignments == null) {
+          ""
+        } else {
+          val assignments = newPartition.assignments.asScala.map(inner => inner.asScala.toList)
+          // check each broker exists
+          val unknownBrokers = assignments.flatten.toSet -- zkUtils.getAllBrokersInCluster.map(broker => broker.id)
+          if (unknownBrokers.nonEmpty) {
+            throw new InvalidReplicaAssignmentException(s"Unknown broker(s) in replica assignment: ${unknownBrokers.mkString(", ")}.")
+          }
+          if (assignments.size != numPartitionsIncrement) {
+            throw new InvalidRequestException(s"Increasing the number of partitions by $numPartitionsIncrement but ${assignments.size} assignments provided.")
+          }
+          "," * (oldNumPartitions) + assignments.map(_.mkString(":")).mkString(",")
```

Comment:
This is not ideal. It's better to have the building block method accept an assignment in a structured form (i.e. not a string).

Reply:
This is difficult to refactor: Currently when invoking `kafka-topics.sh --alter --partitiions ... --replica-assignment ...` we expect a complete replica assignment, but we ignore (not even validating) that part of the replica assignment string which is for the existing partitions. If I were to provide a `parseReplicaAssignment()` function and have `addPartitions()` take a `Map[Int, Seq[Int]]` then either `TopicCommand` or `parseReplicaAssignment()` would have to look up the existing number of partitions for itself to know which are the new assignments in the string. But `addPartitions()` also needs this info, so we consult ZK twice for the same data and the first lookup reintroduces the race wrt `zkUtils.pathExists(ZkUtils.ReassignPartitionsPath)`.

I suppose I can make `addPartitions()` take the complete assignment, rather than the assignment just for the replicas being added, but this retains a poor semantics of passing around information that isn't being used for anything, nor validated.

It might be better to leave it as it is currently and change the `TopicCommand` (which, I understand, is being rewritten anyway) to use a different `--new-replica-assignment` option with `--alter`, retaining the existing `--replica-assignment` for use with `--alter`, thus removing the source of the whole "pass around the complete replica map" issue. Note there are other issues with passing around the complete replica map: It makes it look like `kafka-topics.sh` can be used to reassign existing replicas.

wdyt?

Reply:
What if one of the parameters of `addPartitions` was the current replica assignment? That would avoid the double look-up right?

Reply:
That works, but it does mean we have to trust our caller that the existing assignments are indeed the correct ones, because we use them to update ZK. I've implemented it like this for now.